runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalUpcallHandler.cpp

```diff
@@ -0,0 +1,151 @@
+ */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalUpcallHandler_allocateUpcallStub(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9UpcallMetaData *upcallMetaData = NULL;
+	j9object_t mhMetaData = NULL;
+	J9UpcallNativeSignature *nativeSig = NULL;
+	J9UpcallSigType *sigArray = NULL;
+	void *thunkAddr = NULL;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	j9object_t cSigStrs = J9_JNI_UNWRAP_REFERENCE(currentThread->sp);
+	/* the last element of the array is the signature of return type */
+	U_32 sigCount  = J9INDEXABLEOBJECT_SIZE(currentThread, cSigStrs);
+
+	/* Note: the J9UpcallMetaData pointer will be stored in the generated thunk as data
+	 * in which case it is released only when the generated thunk memory is released
+	 */
+	upcallMetaData = (J9UpcallMetaData *)j9mem_allocate_memory(sizeof(J9UpcallMetaData), J9MEM_CATEGORY_VM_FFI);
+	if (NULL == upcallMetaData) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto done;
+	}
+
+	nativeSig = (J9UpcallNativeSignature *)j9mem_allocate_memory(sizeof(J9UpcallNativeSignature), J9MEM_CATEGORY_VM_FFI);
+	if (NULL == nativeSig) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto freeAllMemoryThenExit;
+	}
+
+	sigArray = (J9UpcallSigType *)j9mem_allocate_memory(sizeof(J9UpcallSigType) * sigCount, J9MEM_CATEGORY_VM_FFI);
+	if (NULL == sigArray) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto freeAllMemoryThenExit;
+	}
+
+	for (U_32 sigIndex = 0; sigIndex < sigCount; sigIndex++) {
+		j9object_t sigStrObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, cSigStrs, sigIndex);
+		char sigBuffer[J9VM_NATIVE_SIGNATURE_STRING_LENGTH] = {0};
+		/* The simplified signature string in cSig for parameter/return type is converted at java level.
+		 * e.g.
+		 * "4#I" represents a 4-byte integer
+		 * "20#[5:I]" represents a 20-byte struct for {int a[5]}
+		 * "16#[I(4)J]" represents a 16-byte struct for {int, padding(4 bytes), long}
+		 *
+		 * Note: the last element is the signature string for the return type.
+		 */
+		char *cSig = copyStringToUTF8WithMemAlloc(currentThread, sigStrObject,
+				J9_STR_NULL_TERMINATE_RESULT, "", 0, sigBuffer, sizeof(sigBuffer), NULL);
+		LayoutFFITypeHelpers::encodeUpcallSignature(cSig, &sigArray[sigIndex]);
+		if (cSig != sigBuffer) {
+			j9mem_free_memory(cSig);
+		}
+	}
+	nativeSig->numSigs = sigCount;
+	nativeSig->sigArray  = sigArray;
+
+	/* Set the fields of the J9UpcallMetaData struct for the thunk generation */
+	upcallMetaData->vm = vm;
+	upcallMetaData->nativeFuncSignature = nativeSig;
+	mhMetaData = J9_JNI_UNWRAP_REFERENCE(currentThread->sp + 1);
+	upcallMetaData->mhMetaData = j9jni_createGlobalRef((JNIEnv*)currentThread, mhMetaData, false);
```

Comment:
This can fail.

Reply:
A null check is added for `upcallMetaData->mhMetaData`.